File: sdk/python/benchmarks/client_benchmark.py
#!/usr/bin/env python3
"""Reproducible external-server benchmark for the Python GlyphaStore SDK."""

from __future__ import annotations

import argparse
import asyncio
import statistics
import sys
import threading
from pathlib import Path
from time import perf_counter
from typing import Callable

PACKAGE_ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(PACKAGE_ROOT / "src"))

from glyphastore import (  # noqa: E402
    AsyncClient,
    Client,
    ClientConfig,
    PipelineOpcode,
    PipelineRequest,
    PipelineResponse,
    __version__,
)


def material(
    operations: int,
    workers: int,
    pipeline: int,
    owner_for: Callable[[bytes], int],
) -> list[list[list[PipelineRequest]]]:
    quotas = [
        operations // workers + (1 if worker < operations % workers else 0)
        for worker in range(workers)
    ]
    requests: list[list[PipelineRequest]] = [[] for _ in range(workers)]
    candidate = 0
    while any(quotas):
        key = f"python-bench-{candidate:012d}".encode()
        owner = owner_for(key)
        if quotas[owner]:
            value = bytes([candidate & 0xFF]) * 64
            requests[owner].append(PipelineRequest(PipelineOpcode.PUT, key, value))
            requests[owner].append(PipelineRequest(PipelineOpcode.GET, key))
            quotas[owner] -= 1
        candidate += 1
    batch_frames = pipeline * 2
    return [
        [
            worker_requests[offset : offset + batch_frames]
            for offset in range(0, len(worker_requests), batch_frames)
        ]
        for worker_requests in requests
    ]


def batch_rounds(
    batches: list[list[list[PipelineRequest]]],
) -> list[list[PipelineRequest]]:
    rounds: list[list[PipelineRequest]] = []
    for offset in range(max(map(len, batches), default=0)):
        rounds.append(
            [
                request
                for worker_batches in batches
                if offset < len(worker_batches)
                for request in worker_batches[offset]
            ]
        )
    return rounds


def validate_responses(batch: list[PipelineRequest], responses: list[PipelineResponse]) -> None:
    if len(responses) != len(batch) or not all(response.succeeded for response in responses):
        raise RuntimeError("pipeline request failed")
    for index in range(1, len(batch), 2):
        if responses[index].value != bytes(batch[index - 1].value):
            raise RuntimeError("pipeline GET value mismatch")


def run_pipeline_concurrent(client: Client, batches: list[list[list[PipelineRequest]]]) -> float:
    barrier = threading.Barrier(len(batches) + 1)
    failures: list[str] = []

    def run(worker_batches: list[list[PipelineRequest]]) -> None:
        barrier.wait()
        for batch in worker_batches:
            responses = client.execute_pipeline(batch)
            if len(responses) != len(batch) or not all(
                response.succeeded for response in responses
            ):
                failures.append("pipeline request failed")
                return
            for index in range(1, len(batch), 2):
                if responses[index].value != bytes(batch[index - 1].value):
                    failures.append("pipeline GET value mismatch")
                    return

    threads = [threading.Thread(target=run, args=(worker_batches,)) for worker_batches in batches]
    for thread in threads:
        thread.start()
    barrier.wait()
    started = perf_counter()
    for thread in threads:
        thread.join()
    elapsed = perf_counter() - started
    if failures:
        raise RuntimeError(failures[0])
    return elapsed


def run_pipeline_sequential(client: Client, batches: list[list[list[PipelineRequest]]]) -> float:
    started = perf_counter()
    for worker_batches in batches:
        for batch in worker_batches:
            responses = client.execute_pipeline(batch)
            if len(responses) != len(batch) or not all(
                response.succeeded for response in responses
            ):
                raise RuntimeError("pipeline request failed")
            for index in range(1, len(batch), 2):
                if responses[index].value != bytes(batch[index - 1].value):
                    raise RuntimeError("pipeline GET value mismatch")
    return perf_counter() - started


def run_batch(client: Client, batches: list[list[list[PipelineRequest]]]) -> float:
    started = perf_counter()
    for batch in batch_rounds(batches):
        responses = client.execute_batch(batch)
        validate_responses(batch, responses)
    return perf_counter() - started


async def run_pipeline_async(
    client: AsyncClient, batches: list[list[list[PipelineRequest]]]
) -> float:
    async def run(worker_batches: list[list[PipelineRequest]]) -> None:
        for batch in worker_batches:
            responses = await client.execute_pipeline(batch)
            if len(responses) != len(batch) or not all(
                response.succeeded for response in responses
            ):
                raise RuntimeError("pipeline request failed")
            for index in range(1, len(batch), 2):
                if responses[index].value != bytes(batch[index - 1].value):
                    raise RuntimeError("pipeline GET value mismatch")

    started = perf_counter()
    await asyncio.gather(*(run(worker_batches) for worker_batches in batches))
    return perf_counter() - started


async def run_batch_async(client: AsyncClient, batches: list[list[list[PipelineRequest]]]) -> float:
    started = perf_counter()
    for batch in batch_rounds(batches):
        responses = await client.execute_batch(batch)
        validate_responses(batch, responses)
    return perf_counter() - started


def report(
    *,
    name: str,
    runtime: str,
    execution: str,
    version: str,
    workers: int,
    pipeline: int,
    operation_count: int,
    samples: list[float],
) -> None:
    rates = [operation_count / sample for sample in samples]
    print("# glyphastore Python client benchmark")
    print(
        f"# sdk_version={version} runtime={runtime} execution={execution} "
        f"workers={workers} pipeline_pairs={pipeline} operations={operation_count}"
    )
    print(
        f"name={name} sdk_version={version} runtime={runtime} execution={execution} "
        f"workers={workers} pipeline_pairs={pipeline} operations={operation_count} "
        f"samples={len(samples)} median_seconds={statistics.median(samples):.9f} "
        f"min_seconds={min(samples):.9f} max_seconds={max(samples):.9f} "
        f"median_ops_per_second={statistics.median(rates):.3f} "
        f"min_ops_per_second={min(rates):.3f} max_ops_per_second={max(rates):.3f}"
    )


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, required=True)
    parser.add_argument("--workers", type=int, default=4)
    parser.add_argument("--ops", type=int, default=200_000, help="PUT/GET pair count")
    parser.add_argument("--pipeline", type=int, default=64, help="PUT/GET pairs per batch")
    parser.add_argument("--warmup", type=int, default=1)
    parser.add_argument("--repeats", type=int, default=7)
    parser.add_argument("--runtime", choices=("sync", "async"), default="sync")
    parser.add_argument(
        "--execution",
        choices=("concurrent", "sequential", "batch"),
        default="concurrent",
        help=(
            "concurrent uses one thread/task per Worker; sequential drains Workers in order; "
            "batch delegates Worker grouping to execute_batch"
        ),
    )
    args = parser.parse_args()
    if min(args.workers, args.ops, args.pipeline, args.repeats) <= 0 or args.warmup < 0:
        parser.error("numeric options must be positive (warmup may be zero)")
    if args.runtime == "async" and args.execution == "sequential":
        # Async sequential is still one event loop; keep the flag for report parity.
        pass

    operation_count = args.ops * 2
    config = ClientConfig(host=args.host, port=args.port)

    if args.runtime == "sync":
        runners = {
            "concurrent": run_pipeline_concurrent,
            "sequential": run_pipeline_sequential,
            "batch": run_batch,
        }
        runner = runners[args.execution]
        with Client.connect(config) as client:
            if client.worker_count != args.workers:
                parser.error("server Worker count does not match --workers")
            batches = material(args.ops, args.workers, args.pipeline, client.worker_for)
            for _ in range(args.warmup):
                runner(client, batches)
            samples = [runner(client, batches) for _ in range(args.repeats)]
        name = (
            "python_client_batch_read_after_write"
            if args.execution == "batch"
            else "python_client_pipeline_read_after_write"
        )
    else:

        async def measure() -> list[float]:
            async with await AsyncClient.connect(config) as client:
                if client.worker_count != args.workers:
                    raise SystemExit("server Worker count does not match --workers")
                batches = material(args.ops, args.workers, args.pipeline, client.worker_for)
                runner = run_batch_async if args.execution == "batch" else run_pipeline_async
                for _ in range(args.warmup):
                    await runner(client, batches)
                return [await runner(client, batches) for _ in range(args.repeats)]

        samples = asyncio.run(measure())
        name = (
            "python_async_client_batch_read_after_write"
            if args.execution == "batch"
            else "python_async_client_pipeline_read_after_write"
        )

    report(
        name=name,
        runtime=args.runtime,
        execution=args.execution
        if args.execution == "batch"
        else (args.execution if args.runtime == "sync" else "concurrent"),
        version=__version__,
        workers=args.workers,
        pipeline=args.pipeline,
        operation_count=operation_count,
        samples=samples,
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/client_benchmark.cpp
#include "glyphastore/client/client.hpp"

#include <algorithm>
#include <atomic>
#include <barrier>
#include <charconv>
#include <chrono>
#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <iostream>
#include <limits>
#include <span>
#include <string>
#include <string_view>
#include <system_error>
#include <thread>
#include <vector>

namespace {

using Clock = std::chrono::steady_clock;
using PipelineRequest = glyphastore::client::PipelineRequest;

struct Options {
    std::string host{"127.0.0.1"};
    std::uint16_t port{};
    std::size_t workers{4};
    std::size_t operations{200'000};
    std::size_t pipeline{64};
    std::size_t warmup{1};
    std::size_t repeats{7};
    std::string execution{"concurrent"};
};

struct Workload {
    std::vector<std::vector<std::string>> keys;
    std::vector<std::vector<std::string>> values;
    std::vector<std::vector<std::vector<PipelineRequest>>> batches;
};

[[nodiscard]] auto as_bytes(const std::string_view value) noexcept -> std::span<const std::byte> {
    return {reinterpret_cast<const std::byte*>(value.data()), value.size()};
}

[[noreturn]] void fail(const std::string_view message) {
    std::cerr << "benchmark error: " << message << '\n';
    std::exit(1);
}

[[nodiscard]] auto parse_size(const std::string_view text, const std::string_view option) -> std::size_t {
    std::size_t value{};
    const auto [end, error] = std::from_chars(text.data(), text.data() + text.size(), value);
    if (error != std::errc{} || end != text.data() + text.size() || value == 0) {
        fail(std::string{option} + " must be a positive integer");
    }
    return value;
}

[[nodiscard]] auto parse_options(const int argc, char** argv) -> Options {
    Options options;
    for (int index = 1; index < argc; ++index) {
        const std::string_view option{argv[index]};
        if (option == "--help") {
            std::cout << "usage: glyphastore_client_benchmark --port N [--host HOST] [--workers N] "
                         "[--ops N] [--pipeline N] [--warmup N] [--repeats N] "
                         "[--execution concurrent|sequential|batch]\n";
            std::exit(0);
        }
        if (index + 1 >= argc) {
            fail(std::string{option} + " requires a value");
        }
        const std::string_view value{argv[++index]};
        if (option == "--host") {
            options.host = value;
        } else if (option == "--port") {
            const auto parsed = parse_size(value, option);
            if (parsed > std::numeric_limits<std::uint16_t>::max()) {
                fail("--port is outside the TCP port range");
            }
            options.port = static_cast<std::uint16_t>(parsed);
        } else if (option == "--workers") {
            options.workers = parse_size(value, option);
        } else if (option == "--ops") {
            options.operations = parse_size(value, option);
        } else if (option == "--pipeline") {
            options.pipeline = parse_size(value, option);
        } else if (option == "--warmup") {
            options.warmup = value == "0" ? 0 : parse_size(value, option);
        } else if (option == "--repeats") {
            options.repeats = parse_size(value, option);
        } else if (option == "--execution") {
            options.execution = value;
        } else {
            fail(std::string{"unknown option: "} + std::string{option});
        }
    }
    if (options.port == 0) {
        fail("--port is required");
    }
    if (options.execution != "concurrent" && options.execution != "sequential" &&
        options.execution != "batch") {
        fail("--execution must be concurrent, sequential, or batch");
    }
    if (options.pipeline > std::numeric_limits<std::size_t>::max() / 2U) {
        fail("--pipeline is too large");
    }
    if (options.operations > std::numeric_limits<std::size_t>::max() / 2U) {
        fail("--ops is too large");
    }
    if (options.workers >= static_cast<std::size_t>(std::numeric_limits<std::ptrdiff_t>::max())) {
        fail("--workers is too large");
    }
    return options;
}

[[nodiscard]] auto make_workload(glyphastore::client::Client& client, const Options& options) -> Workload {
    Workload workload;
    workload.keys.resize(options.workers);
    workload.values.resize(options.workers);
    std::vector<std::size_t> quotas(options.workers, options.operations / options.workers);
    for (std::size_t worker = 0; worker < options.operations % options.workers; ++worker) {
        ++quotas[worker];
    }
    for (std::size_t worker = 0; worker < options.workers; ++worker) {
        workload.keys[worker].reserve(quotas[worker]);
        workload.values[worker].reserve(quotas[worker]);
    }

    std::size_t remaining = options.operations;
    for (std::size_t candidate = 0; remaining != 0; ++candidate) {
        auto key = std::string{"cpp-bench-"} + std::to_string(candidate);
        const auto worker = client.worker_for(key);
        if (worker >= quotas.size() || quotas[worker] == 0) {
            continue;
        }
        workload.keys[worker].push_back(std::move(key));
        workload.values[worker].emplace_back(64, static_cast<char>(candidate & 0xFFU));
        --quotas[worker];
        --remaining;
    }

    workload.batches.resize(options.workers);
    const auto pairs_per_batch = options.pipeline;
    for (std::size_t worker = 0; worker < options.workers; ++worker) {
        auto& worker_batches = workload.batches[worker];
        const auto pair_count = workload.keys[worker].size();
        worker_batches.reserve(pair_count / pairs_per_batch + (pair_count % pairs_per_batch == 0 ? 0U : 1U));
        for (std::size_t begin = 0; begin < pair_count; begin += pairs_per_batch) {
            const auto end = std::min(pair_count, begin + pairs_per_batch);
            auto& batch = worker_batches.emplace_back();
            batch.reserve((end - begin) * 2U);
            for (std::size_t index = begin; index < end; ++index) {
                batch.push_back({.opcode = glyphastore::client::PipelineOpcode::put,
                                 .key = as_bytes(workload.keys[worker][index]),
                                 .value = as_bytes(workload.values[worker][index])});
                batch.push_back({.opcode = glyphastore::client::PipelineOpcode::get,
                                 .key = as_bytes(workload.keys[worker][index])});
            }
        }
    }
    return workload;
}

[[nodiscard]] auto validate(const std::span<const PipelineRequest> requests,
                            const std::vector<glyphastore::client::PipelineResponse>& responses) -> bool {
    if (responses.size() != requests.size()) {
        return false;
    }
    for (std::size_t index = 0; index < requests.size(); ++index) {
        if (!responses[index].succeeded()) {
            return false;
        }
        if (requests[index].opcode == glyphastore::client::PipelineOpcode::get &&
            !std::ranges::equal(responses[index].value, requests[index - 1U].value)) {
            return false;
        }
    }
    return true;
}

[[nodiscard]] auto execute_pipeline(glyphastore::client::Client& client,
                                    const std::vector<PipelineRequest>& requests) -> bool {
    auto responses = client.execute_pipeline(requests);
    return responses && validate(requests, *responses);
}

[[nodiscard]] auto run_sequential(glyphastore::client::Client& client, const Workload& workload) -> double {
    const auto started = Clock::now();
    for (const auto& worker_batches : workload.batches) {
        for (const auto& batch : worker_batches) {
            if (!execute_pipeline(client, batch)) {
                fail("pipeline request failed");
            }
        }
    }
    return std::chrono::duration<double>{Clock::now() - started}.count();
}

[[nodiscard]] auto run_concurrent(glyphastore::client::Client& client, const Workload& workload) -> double {
    std::barrier start{static_cast<std::ptrdiff_t>(workload.batches.size() + 1U)};
    std::atomic_bool failed{};
    std::vector<std::thread> threads;
    threads.reserve(workload.batches.size());
    for (const auto& worker_batches : workload.batches) {
        threads.emplace_back([&client, &start, &failed, &worker_batches] {
            start.arrive_and_wait();
            for (const auto& batch : worker_batches) {
                if (!execute_pipeline(client, batch)) {
                    failed.store(true, std::memory_order_relaxed);
                    return;
                }
            }
        });
    }
    start.arrive_and_wait();
    const auto started = Clock::now();
    for (auto& thread : threads) {
        thread.join();
    }
    if (failed.load(std::memory_order_relaxed)) {
        fail("concurrent pipeline request failed");
    }
    return std::chrono::duration<double>{Clock::now() - started}.count();
}

[[nodiscard]] auto run_batch(glyphastore::client::Client& client, const Workload& workload) -> double {
    std::size_t rounds{};
    for (const auto& worker_batches : workload.batches) {
        rounds = std::max(rounds, worker_batches.size());
    }
    const auto started = Clock::now();
    for (std::size_t round = 0; round < rounds; ++round) {
        std::vector<PipelineRequest> requests;
        for (const auto& worker_batches : workload.batches) {
            if (round < worker_batches.size()) {
                requests.insert(requests.end(), worker_batches[round].begin(), worker_batches[round].end());
            }
        }
        auto responses = client.execute_batch(requests);
        if (!responses || !validate(requests, *responses)) {
            fail("batch request failed");
        }
    }
    return std::chrono::duration<double>{Clock::now() - started}.count();
}

[[nodiscard]] auto median(std::vector<double> values) -> double {
    std::ranges::sort(values);
    const auto middle = values.size() / 2U;
    return values.size() % 2U != 0 ? values[middle] : (values[middle - 1U] + values[middle]) / 2.0;
}

} // namespace

int main(const int argc, char** argv) {
    const auto options = parse_options(argc, argv);
    auto connected = glyphastore::client::Client::connect({
        .host = options.host,
        .port = options.port,
        .maximum_pipeline_requests = options.pipeline * 2U,
    });
    if (!connected) {
        fail(connected.error().message);
    }
    auto client = std::move(*connected);
    if (client.worker_count() != options.workers) {
        fail("server Worker count does not match --workers");
    }
    const auto workload = make_workload(client, options);
    const auto run = [&] {
        if (options.execution == "sequential") {
            return run_sequential(client, workload);
        }
        if (options.execution == "batch") {
            return run_batch(client, workload);
        }
        return run_concurrent(client, workload);
    };
    for (std::size_t index = 0; index < options.warmup; ++index) {
        (void)run();
    }
    std::vector<double> samples;
    samples.reserve(options.repeats);
    for (std::size_t index = 0; index < options.repeats; ++index) {
        samples.push_back(run());
    }
    std::vector<double> rates;
    rates.reserve(samples.size());
    const auto operation_count = options.operations * 2U;
    for (const auto sample : samples) {
        rates.push_back(static_cast<double>(operation_count) / sample);
    }
    const auto name = options.execution == "batch" ? "cpp_client_batch_read_after_write"
                                                   : "cpp_client_pipeline_read_after_write";
    std::cout << "# glyphastore C++ client benchmark\n"
              << "# sdk_version=" << GLYPHASTORE_SDK_VERSION
              << " runtime=native execution=" << options.execution << " workers=" << options.workers
              << " pipeline_pairs=" << options.pipeline << " operations=" << operation_count << '\n'
              << "name=" << name << " sdk_version=" << GLYPHASTORE_SDK_VERSION
              << " runtime=native execution=" << options.execution << " workers=" << options.workers
              << " pipeline_pairs=" << options.pipeline << " operations=" << operation_count
              << " samples=" << samples.size() << " median_seconds=" << median(samples)
              << " min_seconds=" << *std::ranges::min_element(samples)
              << " max_seconds=" << *std::ranges::max_element(samples)
              << " median_ops_per_second=" << median(rates)
              << " min_ops_per_second=" << *std::ranges::min_element(rates)
              << " max_ops_per_second=" << *std::ranges::max_element(rates) << '\n';
}
